sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -960,22 +959,47 @@ public void UpdateConcurrency(int maxConcurrentCalls)
             }
         }
 
-        internal void UpdateConcurrency(int maxConcurrentSessions, int maxConcurrentCallsPerSession)
+        internal void UpdateConcurrency(int maxConcurrentSessions, int maxConcurrentCallsPerSession, int? maxConcurrentCallsAcrossAllSessions = default)
         {
             Argument.AssertAtLeast(maxConcurrentSessions, 1, nameof(maxConcurrentSessions));
             Argument.AssertAtLeast(maxConcurrentCallsPerSession, 1, nameof(maxConcurrentCallsPerSession));
 
+            if (maxConcurrentCallsAcrossAllSessions.HasValue)
+            {
+                Argument.AssertAtLeast(maxConcurrentCallsAcrossAllSessions.Value, 1, nameof(maxConcurrentCallsAcrossAllSessions));
+            }
+
             lock (_maxConcurrencySyncLock)
             {
-                _maxConcurrentCalls = _sessionIds.Length > 0
-                    ? Math.Min(_sessionIds.Length, maxConcurrentSessions)
-                    : maxConcurrentSessions * maxConcurrentCallsPerSession;
                 _maxConcurrentSessions = maxConcurrentSessions;
                 _maxConcurrentCallsPerSession = maxConcurrentCallsPerSession;
+
+                SetMaxConcurrentCallsAcrossSessions(maxConcurrentCallsAcrossAllSessions);
                 WakeLoop();
             }
         }
 
+        private void SetMaxConcurrentCallsAcrossSessions(int? maxConcurrentCallsAcrossAllSessions)
+        {
+            int calculatedMaxConcurrentCalls = _sessionIds.Length > 0
+                ? Math.Min(_sessionIds.Length, _maxConcurrentSessions)
+                : _maxConcurrentSessions * _maxConcurrentCallsPerSession;
```

Comment:
If the caller were to use int.MaxValue for MaxConcurrentCallsPerSession to signal no constraint there, this will result in integer overflow. So we need a safe way to allow this to be specified. Perhaps rather than int.MaxValue, I should just pass in some large value like 500 that I know won't be exceeded? E.g. 500 is the default maximum per-function concurrency limit.

Reply:
Good catch - yes using the max per-function value makes sense.